### PR TITLE
Layout alloc primitives

### PR DESCRIPTION
**Pull request by Britefury**
Implementation of space allocation primitive for layout.

### CODE REVIEW

File: src/lib/lspace/layout/lalloc.rs [hunk 1]
```diff
@@ -0,0 +1,768 @@
...
+    /// `region_pos` : the position of the region in the parent space; most elements that have
+    /// multiple children will want to give them different positions
+    /// `region_size` : the size of the region
+    /// `region_ref` : the reference point of the region (optional)
+    pub fn alloc_from_region(&mut self, child_req: &LReq, region_pos: f64, region_size: f64,
+                             region_ref: Option<f64>) {
+        match child_req.size().before_and_after_ref_opt() {
+            Some((req_before, req_after)) => {
+                // Both the allocation region and the requisition have a reference point
+
```
Comment (ltratt) [inline src/lib/lspace/layout/lalloc.rs]: Random blank line?

File: src/lib/lspace/layout/lalloc.rs [hunk 2]
```diff
@@ -0,0 +1,768 @@
...
+    /// 'child_reqs' : a vector containing an `LReq` for each child element
+    /// 'child_allocs' : a vector containing an `LAlloc` for each child element
+    /// 'region_req' : the `LReq` for the region within the parent element
+    /// `region_pos`, `region_size`, `region_ref` : the position, size and optional reference
+    /// point that describe the region in which the child elements are to be positioned
+    /// `space_between` : the amount of space to insert between child elements
+    /// `ref_point_index` : an optional index that identifies a child element that is to have its
+    /// reference point aligned with that of the region
+    pub fn alloc_linear(child_reqs: &Vec<&LReq>, child_allocs: &mut Vec<&mut LAlloc>,
+                        region_req: &LReq,
```
Comment (ltratt) [inline src/lib/lspace/layout/lalloc.rs]: This argument looks all lonely on its own line.

File: src/lib/lspace/layout/lalloc.rs [hunk 3]
```diff
@@ -0,0 +1,768 @@
...
+                        // the required reference point aligns with the allocated reference point.
+                        // Assuming that the region requisition has its reference point
+                        // aligned with that of the correct child, this should be fine...
+                        let before_offset = fast_max(
+                            region_before_ref - region_req.size().before_ref(), 0.0);
+                        LAlloc::_alloc_children_linear(n, &child_reqs[..],
+                            &mut child_allocs[..],
+                            before_offset + region_pos, space_between);
+                    },
+                    &LFlex::Flex{..} => {
```
Comment (ltratt) [inline src/lib/lspace/layout/lalloc.rs]: Isn't this equivalent to `&LFlex::Flex`?

File: src/lib/lspace/layout/lalloc.rs [hunk 4]
```diff
@@ -0,0 +1,768 @@
...
+            (_, &LNatSize::Empty, None) => {
+                // Ensure children are all empty
+                for (i, ch) in child_reqs.iter().enumerate() {
+                    if ch.size() != &LNatSize::Empty {
+                        panic!("Empty parent requisition, non-empty child requisiton at \
+                               index {}", i);
+                    }
+                }
+
+                // Nothing to do
```
Comment (ltratt) [inline src/lib/lspace/layout/lalloc.rs]: Not quite sure why this comment is here?

**Comment by ltratt** [comment]
Apart from the one very, very long `match` statement -- where I almost wonder if it should be split into separate functions? -- this looks in good shape.

File: src/lib/lspace/layout/lalloc.rs (hunk 3 above)
Comment (Britefury) [inline src/lib/lspace/layout/lalloc.rs]: Fraid not. When an enum value has structure members you have to mention this otherwise rust complains.

Comment (ltratt) [inline src/lib/lspace/layout/lalloc.rs]: Ah, yes, of course!

File: src/lib/lspace/layout/lalloc.rs [hunk 5]
```diff
@@ -0,0 +1,805 @@
...
+            let creq = child_reqs[i];
+            let csz = creq.size().size();
+            child_allocs[i].alloc_from_region(creq, pos, csz, creq.size().before_ref_opt());
+            pos = pos + csz + space_between;
+        }
+        return pos;
+    }
+
+    fn _alloc_children_linear_flex(n: usize,
+                                   child_reqs: &[&LReq], child_allocs: &mut [&mut LAlloc],
```
Comment (ltratt) [inline src/lib/lspace/layout/lalloc.rs]: Our of curiosity, why have these as (in effect) static methods on the struct? They don't seem to be directly related to it (though they are tangentially related).

**Comment by ltratt** [comment]
I just realised: there are lots of things starting with `_`. I assume this is bringing the Python concept of "private by convention" to Rust? If so, I'm not sure this is a Rusty thing to do. Given that Rust is private by default, the convention seems to be generally not use `_` as a prefix (except in a few cases I've seen in the standard library where it's used to avoid a name clash).

**Comment by Britefury** [comment]
The `_` function name prefixes are gone and that large match statement has been reduced in size by moving functionality into helper functions. The static helper functions no longer reside with `LAlloc`.

This is still work-in-progress though; unit tests for `alloc_linear` need to be written.

**Comment by ltratt** [comment]
Looks good.

I have one possibly dumb question: why does lspace mix `f32`s and `f64`s? For example, `shrink` returns an `f64` but `stretch` returns an `f32`. This then leads to casting, and I can't help wondering if that's going to lead to some subtle bugs at some point?

**Comment by Britefury** [comment]
Absolute positions and sizes use `f64` for the added precision; `shrink` is the amount (probably in pixels) by which a box can shrink if less space than its natural size is available. `stretch` is different; if additional space beyond the natural size is available for an element, it is apportioned to elements proportional to the value of `stretch`. It is therefore unlikely to require the same precision as `shrink`. As a consequence I decided to use `f32` in order to save a little space.

**Comment by ltratt** [comment]
Is there ever any chance of a f64 being downcast to f32?

**Comment by Britefury** [comment]
Not AFAIK. I think rust prevents casting between `f32` and `f64` without explicitly requesting it.

**Comment by ltratt** [comment]
I can see some upcasts, but wasn't sure about downcasts.

**Comment by Britefury** [comment]
Ready for review.

**Comment by ltratt** [comment]
Which bits shall I review? I guess only some of the commits, not all of them?

**Comment by Britefury** [comment]
Hmmm. The first (`4398914`) has a refactoring that you may be curious about; using iterators rather than `Vec`. The last 4 are all bug fixes and tests for the `LAlloc::alloc_linear()` method. I guess `c174ee4` since it finishes the unit testing off.

**Comment by ltratt** [comment]
Ah, so you mean all the commits from 4398 onwards?

**Comment by Britefury** [comment]
I think 4398 and c174 are the most important of those 5. You could take a look at the 3 in between if you like.

**Comment by ltratt** [comment]
OK, I've made some minor comments. Since 4fa322b didn't pass tests, I think that should be squashed away entirely.

I'm curious about the various `asserts` (well, `assert_almost_eq`): are these definitely worth running at all points? Or could they be `debug_assert`s that are optimised away with `--release`?

**Comment by Britefury** [comment]
The caps comments (e.g. 'FIXED') have been changed.
Commented code is gone.
The 1.0e-6 is a constant.
The asserts in run-time code have been changed to `debug_assert`.
The `assert_almost_eq` are all used in tests.
5 commits squashed into 1, plus a comment fix commit.

Ready for review.

**Comment by ltratt** [comment]
Let's squash the "comment fixes" commit... then I assume this is ready to be merged?

**Comment by Britefury** [comment]
Okay, squashed. Ready for merge.

**Comment by ltratt** [comment]
Looks good!

**Pull request merged**